src/app/core/services/rundown-state.service.ts

```diff
@@ -1,114 +1,186 @@
 import { Part } from '../models/part';
-
-interface RundownSubjectInfo {
-  rundownId: string
-  subject: BehaviorSubject<Rundown | undefined>,
-  subscriberCount: number
-}
+import { RundownEventObserver } from './events/rundown-event-observer.service'
+import { Unsubscribe } from './events/event-observer.interface'
+import { ManagedSubscription } from './managed-subscription.service'
+import { ConnectionStatusObserver } from './events/connection-status-observer.service'
 
 @Injectable()
-export class RundownStateService {
-  private subjectInfoMap: Map<string, RundownSubjectInfo> = new Map()
+export class RundownStateService implements OnDestroy {
+  private readonly rundownSubjects: Map<string, BehaviorSubject<Rundown>> = new Map()
+  private unsubscribeFromEvents: Unsubscribe
 
   constructor(
-    private rundownService: RundownService,
-    private rundownEventService: RundownEventService
-  ) { }
-
-  public subscribeToRundown(rundownId: string, callback: (rundown: Rundown | undefined) => void): () => void {
-    let subjectInfo: RundownSubjectInfo | undefined = this.subjectInfoMap.get(rundownId)
-    if (!subjectInfo) {
-      subjectInfo = { rundownId, subject: new BehaviorSubject<Rundown | undefined>(undefined), subscriberCount: 0 }
-      this.subjectInfoMap.set(rundownId, subjectInfo)
-      this.fetchRundown(rundownId)
-      this.startListeningForRundownEvent(rundownId)
-    }
-    subjectInfo.subscriberCount += 1
-    const subscription = subjectInfo.subject.subscribe(callback)
-    return () => this.unsubscribeFromSubject(subjectInfo!, subscription)
+      private readonly rundownService: HttpRundownService,
+      private readonly rundownEventObserver: RundownEventObserver,
+      private readonly connectionStatusObserver: ConnectionStatusObserver
+  ) {
+    this.registerEventConsumers()
+  }
+
+  private registerEventConsumers(): void {
+    const unsubscribeFromConnectionStatusEvents = this.registerConnectionStatusConsumers()
+    const unsubscribesFromRundownEvents = this.registerRundownEventConsumers()
+    this.unsubscribeFromEvents = () => [
+        ...unsubscribesFromRundownEvents,
+        ...unsubscribeFromConnectionStatusEvents
+      ].forEach(unsubscribe => unsubscribe())
+  }
+
+  private registerConnectionStatusConsumers(): Unsubscribe[] {
+    return [
+      this.connectionStatusObserver.subscribeToConnectionStatus(this.resetRundownSubjectsIfConnected.bind(this))
+    ]
   }
 
-  private unsubscribeFromSubject(subjectInfo: RundownSubjectInfo, subscription: Subscription) {
-    subscription.unsubscribe()
-    subjectInfo.subscriberCount -= 1
-    if (subjectInfo.subscriberCount === 0) {
-      this.subjectInfoMap.delete(subjectInfo.rundownId)
-      this.stopListeningForRundownEvents(subjectInfo.rundownId)
+  private resetRundownSubjectsIfConnected(isConnected: boolean): void {
+    if (!isConnected) {
+      return
+    }
+    for (let [rundownId, rundownSubject] of this.rundownSubjects.entries()) {
```

Comment:
Any reason why the `let` can't be a `const`?

Reply:
Nope, I think it was a WebStorm snippet :D I have updated it to use const.

Reply:
Yes, it's a loop